TestPullVerified fails on Docker master branch

Running the TestPullVerified test (as part of 'make test-integration-cli') fails on the current master branch because the expected string "The image you are pulling has been verified" does not appear.

Here is the output I'm seeing (note the image is already present, so verified message is not shown):
--- FAIL: TestPullVerified (0.81s)
        docker_cli_pull_test.go:71: pulling a verified image failed. expected: The image you are pulling has been verified
                got: latest: Pulling from hello-world
                511136ea3c5a: Already exists
                31cbccb51277: Already exists
                e45a5af57b00: Already exists
                Status: Image is up to date for hello-world:latest
                , <nil>
FAIL

--- a/Dockerfile
+++ b/Dockerfile
@@ -146,7 +146,7 @@ RUN ln -sfv $PWD/.bashrc ~/.bashrc
 COPY contrib/download-frozen-image.sh /go/src/github.com/docker/docker/contrib/
 RUN ./contrib/download-frozen-image.sh /docker-frozen-images \
 	busybox:latest@4986bf8c15363d1c5d15512d5266f8777bfba4974ac56e3270e7760f6f0a8125 \
-	hello-world:latest@e45a5af57b00862e5ef5782a9925979a02ba2b12dff832fd0991335f4a11e5c5
+	hello-world:frozen@e45a5af57b00862e5ef5782a9925979a02ba2b12dff832fd0991335f4a11e5c5
 # see also "hack/make/.ensure-frozen-images" (which needs to be updated any time this list is)
 
 # Install man page generator

--- a/graph/pull.go
+++ b/graph/pull.go
@@ -532,7 +532,7 @@ func (s *TagStore) pullV2Tag(eng *engine.Engine, r *registry.Session, out io.Wri
 		}
 	}
 
-	var layersDownloaded bool
+	var tagUpdated bool
 	for i := len(downloads) - 1; i >= 0; i-- {
 		d := &downloads[i]
 		if d.err != nil {
@@ -556,14 +556,27 @@ func (s *TagStore) pullV2Tag(eng *engine.Engine, r *registry.Session, out io.Wri
 				// FIXME: Pool release here for parallel tag pull (ensures any downloads block until fully extracted)
 			}
 			out.Write(sf.FormatProgress(common.TruncateID(d.img.ID), "Pull complete", nil))
-			layersDownloaded = true
+			tagUpdated = true
 		} else {
 			out.Write(sf.FormatProgress(common.TruncateID(d.img.ID), "Already exists", nil))
 		}
 
 	}
 
-	if verified && layersDownloaded {
+	// Check for new tag if no layers downloaded
+	if !tagUpdated {
+		repo, err := s.Get(repoInfo.LocalName)
+		if err != nil {
+			return false, err
+		}
+		if repo != nil {
+			if _, exists := repo[tag]; !exists {
+				tagUpdated = true
+			}
+		}
+	}
+
+	if verified && tagUpdated {
 		out.Write(sf.FormatStatus(repoInfo.CanonicalName+":"+tag, "The image you are pulling has been verified. Important: image verification is a tech preview feature and should not be relied on to provide security."))
 	}
 
@@ -571,5 +584,5 @@ func (s *TagStore) pullV2Tag(eng *engine.Engine, r *registry.Session, out io.Wri
 		return false, err
 	}
 
-	return layersDownloaded, nil
+	return tagUpdated, nil
 }

--- a/hack/make/.ensure-frozen-images
+++ b/hack/make/.ensure-frozen-images
@@ -4,7 +4,7 @@ set -e
 # this list should match roughly what's in the Dockerfile (minus the explicit image IDs, of course)
 images=(
 	busybox:latest
-	hello-world:latest
+	hello-world:frozen
 )
 
 if ! docker inspect "${images[@]}" &> /dev/null; then

--- a/hack/make/.ensure-frozen-images
+++ b/hack/make/.ensure-frozen-images
@@ -11,10 +11,10 @@ if ! docker inspect "${images[@]}" &> /dev/null; then
 	hardCodedDir='/docker-frozen-images'
 	if [ -d "$hardCodedDir" ]; then
 		( set -x; tar -cC "$hardCodedDir" . | docker load )
-	elif [ -e Dockerfile ] && command -v curl > /dev/null; then
-		# testing for "curl" because "download-frozen-image.sh" is built around curl
+	else
 		dir="$DEST/frozen-images"
 		# extract the exact "RUN download-frozen-image.sh" line from the Dockerfile itself for consistency
+		# NOTE: this will fail if either "curl" is not installed or if the Dockerfile is not available/readable
 		awk '
 			$1 == "RUN" && $2 == "./contrib/download-frozen-image.sh" {
 				for (i = 2; i < NF; i++)
@@ -33,11 +33,5 @@ if ! docker inspect "${images[@]}" &> /dev/null; then
 			}
 		' Dockerfile | sh -x
 		( set -x; tar -cC "$dir" . | docker load )
-	else
-		for image in "${images[@]}"; do
-			if ! docker inspect "$image" &> /dev/null; then
-				( set -x; docker pull "$image" )
-			fi
-		done
 	fi
 fi

--- a/integration-cli/docker_cli_build_test.go
+++ b/integration-cli/docker_cli_build_test.go
@@ -5234,7 +5234,7 @@ func TestBuildRUNoneJSON(t *testing.T) {
 	defer deleteAllContainers()
 	defer deleteImages(name)
 
-	ctx, err := fakeContext(`FROM hello-world:latest
+	ctx, err := fakeContext(`FROM hello-world:frozen
 RUN [ "/hello" ]`, map[string]string{})
 	if err != nil {
 		t.Fatal(err)

